source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md

```diff
@@ -0,0 +1,254 @@
+---
+layout: post
+title: "How to avoid @inject thanks to Decorator feature in Nette"
+perex: "I often find <code>@inject</code> being overused in projects I review while mentoring. They often bring less writing, but in exchange they break SOLID principals. Today I will show you solution that will <strong>keep your code both small and clean</strong> - <strong>Decorator feature in Nette</strong>."
+lang: en
+---
+
+As [Derek Simons says](https://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action) says...
+
+## ...Start with "Why"
+
+Why am I writing this article? I try to improve knowledge interoperability between frameworks by finding the common lowest denominator, so it **is easier to understand and use each other**.
+
+I understand every framework has its own specific way how to solve same issues, but it creates *knowledge vendor lock*. So even if I know how to solve some problem in Nette, I cannot do that in Symfony and have to re-learn it again.
+ 
+Today, I will try to agree on setter injection with you. 
+
+
+## @Inject Overuse is Spreading 
+
+This code is common to 80 % Nette applications I came across in last year:
+
+```language-php
+// app/Presenter/ProductPresenter.php
+
+namespace App\Presenter;
+
+final class ProductPresenter extends AbstractBasePresenter
+{
+    /**
+     * @inject
+     * @var ProductRepository
+     */
+    public $productRepository;
+}
+```
+
+Using `@inject` annotations over constructor injection is **fast, short and it just works**.
+
+Ok, why not use it everywhere:
+
+```language-php
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+class ProductRepository
+{
+    /**
+     * @inject
+     * @var Doctrine
+     */
+    public $entityManager;
+}
+```
+
+and 
+
+```language-yaml
+# app/config/config.neon
+
+services:
+    -
+        class: App\Repository\ProductRepository
+            inject: on
+```
+
+## Your Code is Seen as Manual How to Write 
+
+Why? Because "what you see is what you write". New programmer joins the teams, sees this handy `@inject` feature and uses when possible and handy.
+
+Some of you, who already talked about `@inject` method usage already there are some and only few specific places to use it.
+
+## Where to only @inject?
+
+**To prevent constructor hell**. If you meet this term first time, go read [this short explanation](https://phpfashion.com/di-a-predavani-zavislosti#toc-constructor-hell) from David Grudl.
```

Comment:
I actually try not to complain, since I see no added value in it apart relieving emotions.

Rather I try to suggest realistic path to better design, that might lead to solution to your note in the future.